### DISCUSSION TITLE
Estudos com exercícios para aprender na prática.

### DISCUSSION THREAD

**Discussion by synga** [frontendbr/forum#1062] [Discussão]
Fala galera, estou aqui passando por um certo problema com estudos que talvez alguns também estejam passando ou já passaram. Comprar cursos, aprender com videoaulas, YouTube, PluralSight, Udemy entre outros é bom, mas a maioria não tem uma seção de exercícios e se você não pratica acaba estacionando o conhecimento e quando vai usar precisa ficar consultando o Google (to assim com grid de CSS que fiz um curso há 2 meses). E gostaria de saber se vocês tem sites, canais, cursos, algo pra aprender, porém praticando a cada aula ou cada seção. Fiz um cursinho do FreeCodeCamp de JS a uns tempos atrás e foi ótimo, era uma aula explicando algo novo e uma prática. Também utilizo o [repositório de desafios](https://github.com/LFeh/frontend-challenges) do @LFeh que são ótimos de se fazer.

Então conhecem mais algum repositório? Algum curso que tenha bastante prática? Algum site que gere projetos pra codar? Qualquer coisa? Pode ser qualquer coisa relacionada a frontend, com preferencia pela tríade básica do HTML/CSS/JS. Procurei no grupo algo assim, porém achei mais sobre livros e nada mais prático, caso tenha passado batido por alguma issue só jogar ai.

Obrigado.

## EDIT ##
Achei o site que falei, chama [Designercize](https://designercize.com/) e nele você gera projetos nos níveis fácil, médio e difícil, totalmente randômicos, com o que precisa ser feito, para quem e para qual propósito. Ele também tem um cronometro caso você queira tentar fazer dentro de um determinado tempo.

**Comment by kvnol** [comment]
Os cursos do @fdaciuk são ótimos, toda aula tem exercício.

**Comment by leandroruel** [comment]
eu assisto aula da udemy com editor do lado fazendo igualzinho ou adaptando

**Comment by simaodeveloper** [comment]
1 - O ideal é planejar um projeto, mesmo que você não saiba como construí-lo, é necessário saber o que construir.

2 - Depois você quebra ele em partes como (não precisar dominar photoshop, illustrator para construir um projeto simples): 
Idéia -> Brainstorm -> Layout

3 - Definida a estrutura do layout, por mais que seja simples, comece a definir as tarefas para o Front-End, nessa etapa você vai começar a identificar as coisas que vai precisar aprender, e o legal é que tu já vai aplicando.

Exemplo: Preciso estruturar o layout do site com CSS, pesquisei no google algo parecido e ele me apresentou esse site http://pt-br.learnlayout.com/, entendi como se utilizam as propriedades e já fui aplicando no layout que defini anteriormente, ai com o tempo a gente descobre que tem Flexbox e Grid, então, em um proximo projeto eu aplico esses e vou aprendendo, claro tudo dentro dos limites do projeto (Browser Support).

4 - A dica aqui é identificar o que você precisa fazer, uma visão macro de exemplo, preciso criar um fluxo de cadastro, preciso adicionar um calendário. Uma visão micro, eu tenho uma string e preciso transformar em uma lista (Ex: "A B C", bom eu tenho uma string, tem o método split que quebra a string em items informando um **separador**, se eu passar o espaço como **separador**, ele me retorna os dados que eu preciso em uma list :D")

5 - Agora eu preciso adicionar alguma interação, animação, validação, pesquise o que precisa e já aplique, só assim o conhecimento fixa.

Relato Pessoal: Eu estudei VueJS durante 3 meses, construindo aplicações de tutoriais, não absorvi muito conhecimento, agora eu estou criando um projeto e aplicando ele, e ta rolando muitas dúvidas, agora estou aprendendo de verdade, não adianta só estudar mesmo, você esta certíssimo em querer praticar, mas minha dica é crie seus próprios projetos, vai te ajudar muito.

Uma Dica: Começe pelo seu portifólio (Eu tenho que seguir essa dica também aushuahuahs)

**Comment by didiraja** [comment]
Te entendo bem @synga, entrei num rant de estudo recentemente e passei pela mesma coisa. A melhor solução que arrumei foi a mesma que o @simaodeveloper falou, começar a inventar uns projetos da caixola ou fazer que nem eu e prometer ajudar amigos pra te forçar a desenvolver melhor.

**Comment by synga** [comment]
@leandroruel eu faço essa abordagem também, porém também crio um arquivo .txt e vou pausando o vídeo e anotando o mais detalhado possível o que vou aprendendo para aquilo entrar mais fácil na cabeça, mas não vou mentir que no fundo me sinto fazendo um copy/paste por só ir copiando o que esta sendo ensinado, vou tentar variar o código, ver outras aplicações e etc. para poder praticar um pouco mais. Obrigado pela dica, vou sempre deixar um arquivinho separado fora o que já está rolando na aula pra ir praticando variações o que estou aprendendo.

@simaodeveloper nem me fale em portfólio, comecei, fiz um basicão e ta lá hahahahaha eu gosto dessa abordagem de projetos, mas aqui fica meu problema com isso e talvez você saiba me ajudar a contornar melhor, você não acaba focando em fazer o projeto apenas com o que já sabe? Você não acaba deixando de estudar coisas novas por estar exercitando uma forma que já conhece?
Eu vou tentar melhor essa abordagem, porém a cada passo vou ir pesquisando se não tem uma forma melhor de fazer, como nos passos 4 e 5 que você descreveu ai, já vou começar hoje a bolar algum projeto legal pra botar em prática. Obrigado cara.

Alias dei update na minha issue com um site que gera projetinhos pra você desenvolver, alguns são rápidos como um único componente, outros já são projetos completos, é bom pra quem as vezes não consegue ter ideia de algo pra se fazer (como eu).

**Comment by pedromuraki** [comment]
O [CodeFights](https://codefights.com/) é legal pra praticar JS.. Não são projetos, mas sim exercícios pra você resolver usando JS (ou outras linguagens), em que você vai subindo de níveis, ganhando pontos, etc. Tem uns desafios de empresas e desafios de usuários também.. Acho bem legal pra desenvolver lógica, aprender e fixar melhor os métodos JS.

**Comment by mahenrique94** [comment]
Para cursos gosto e recomendo a [Alura](https://www.alura.com.br/) justamente porque é uma das poucas plataformas que tem bastante exercício durante os cursos.

Além disso, tem o [CodeFights](https://codefights.com/) onde você pode praticar qualquer linguagem de sua escolha (tem uma grande variedade), a ideia é que você passe por desafios que vão ficando cada vez mais dificeis e compare sua solução com de outras pessoas.

Também gosto além de realizar os projetos dos cursos que faço, criar os meus próprios, utilizando as mesmas tecnologias.

**Comment by ninetails** [comment]
Sobre sites de desafio, recomendo o [HackerRank](https://www.hackerrank.com/), muita gente usa pra aplicar testes pra seleção. Se não me engano, acertando ou errando acho que dá pra ver a solução depois do problema... não deixa de ser um jeito de aprender praticando. Tem séries que estão formatados muito mais didáticos que se comparados com sites que focam em Coding Challenge.

**Comment by vsppedro** [comment]
Outra opção  para Coding Challenge que acho interessante, além do HackerRank, é o [Codewars](http://www.codewars.com). Pela gamificação e por permitir acesso a resposta dos outros usuários após a solução. Elas sãos listadas por criatividade e melhores práticas. É muito bom ver/estudar o código de desenvolvedores mais experiente.

Agora para desenvolvimento de projetos, um site que me chamou a atenção recentemente é o [Thinkster](https://thinkster.io/). Existem vários tutorials para desenvolvimento de um mesmo sistema. O interessante é que ele permite que você escolha as tecnologias que quer estudar. Exemplo: React + Django. Agora imagine que depois de fazer os tutoriais desses dois você decide que quer ver como seria o comportamento do mesmo só que com Django + AngularJS. O legal é que você só vai precisar pegar o tutorial para AngularJS, e só. A API Django não vai precisar sofrer nenhuma mudança, já está pronta.

O detalhe dessa plataforma é que não lhe da acesso a todo o conteúdo do tutorial de forma free...

Gostei do site [Designercize](https://designercize.com/)! Interessante!

**Comment by diessica** [comment]
se está procurando algo mais mundo real, desenvolver projetos mesmo, postei alguns links para repositórios com ideias de projetos nessa thread: https://github.com/frontendbr/forum/issues/1073#issuecomment-387641827

**Comment by faelribeiro22** [comment]
Tem esse repositório com exemplos de projetos do mundo real bem interessante.
https://github.com/gothinkster/realworld/

**Comment by maghace** [comment]
Estou estudando programação a uns dois meses, de forma auto-didata, mesmo tentando as vezes após algumas aulas tentar ler documentações e aplicar e misturar, testar ideias, sempre da um branco no objetivo/desafio a ser ultrapassado. Esse post é o primeiro sendo um 'noob user' do github, mas já expandiu as ideias de 'o que procurar?', 'onde procurar?' e 'até onde posso ir?'. Espero também pode contribuir em futuras publicações, para os users da comunidade.

**Comment by onlurking** [comment]
Na linha de praticar layouts e CSS tem mais alguns sites:

- [Front-End Challenges Club](https://front-end-challenges.club/archive) do [Andy Bell](https://twitter.com/hankchizljaw). ¹
- [Frontend Mentor](https://www.frontendmentor.io/challenges), com vários designs de telas pra praticar.
- [Frontloops](https://frontloops.io/), que você recebe por e-mail os desafios juntamente com o assets(imagens, etc). ¹

E no esquema de jogos tem o [CSS Battle](https://cssbattle.dev/) aonde você replica os desafios usando o mínimo de CSS possível.

¹ As soluções nesses sites são pagas.

**Comment by Fabio-Pina** [comment]
> Fala galera, estou aqui passando por um certo problema com estudos que talvez alguns também estejam passando ou já passaram. Comprar cursos, aprender com videoaulas, YouTube, PluralSight, Udemy entre outros é bom, mas a maioria não tem uma seção de exercícios e se você não pratica acaba estacionando o conhecimento e quando vai usar precisa ficar consultando o Google (to assim com grid de CSS que fiz um curso há 2 meses). E gostaria de saber se vocês tem sites, canais, cursos, algo pra aprender, porém praticando a cada aula ou cada seção. Fiz um cursinho do FreeCodeCamp de JS a uns tempos atrás e foi ótimo, era uma aula explicando algo novo e uma prática. Também utilizo o [repositório de desafios](https://github.com/LFeh/frontend-challenges) do @LFeh que são ótimos de se fazer.
> 
> Então conhecem mais algum repositório? Algum curso que tenha bastante prática? Algum site que gere projetos pra codar? Qualquer coisa? Pode ser qualquer coisa relacionada a frontend, com preferencia pela tríade básica do HTML/CSS/JS. Procurei no grupo algo assim, porém achei mais sobre livros e nada mais prático, caso tenha passado batido por alguma issue só jogar ai.
> 
> Obrigado.
> ## EDIT
> 
> Achei o site que falei, chama [Designercize](https://designercize.com/) e nele você gera projetos nos níveis fácil, médio e difícil, totalmente randômicos, com o que precisa ser feito, para quem e para qual propósito. Ele também tem um cronometro caso você queira tentar fazer dentro de um determinado tempo.

Olá, não sei se vai lhe ajudar ou se já achou algo nesse sentido pois a postagem é de 2018 mas a W3C Schools tem uns exercícios assim ao final de cada tópico e diz se você acertou ou não. Sobre projetos não entendo ainda. Espero que eu tenha ajudado.

**Comment by Fabio-Pina** [comment]
> @leandroruel eu faço essa abordagem também, porém também crio um arquivo .txt e vou pausando o vídeo e anotando o mais detalhado possível o que vou aprendendo para aquilo entrar mais fácil na cabeça, mas não vou mentir que no fundo me sinto fazendo um copy/paste por só ir copiando o que esta sendo ensinado, vou tentar variar o código, ver outras aplicações e etc. para poder praticar um pouco mais. Obrigado pela dica, vou sempre deixar um arquivinho separado fora o que já está rolando na aula pra ir praticando variações o que estou aprendendo.
> 
> @simaodeveloper nem me fale em portfólio, comecei, fiz um basicão e ta lá hahahahaha eu gosto dessa abordagem de projetos, mas aqui fica meu problema com isso e talvez você saiba me ajudar a contornar melhor, você não acaba focando em fazer o projeto apenas com o que já sabe? Você não acaba deixando de estudar coisas novas por estar exercitando uma forma que já conhece?
> Eu vou tentar melhor essa abordagem, porém a cada passo vou ir pesquisando se não tem uma forma melhor de fazer, como nos passos 4 e 5 que você descreveu ai, já vou começar hoje a bolar algum projeto legal pra botar em prática. Obrigado cara.
> 
> Alias dei update na minha issue com um site que gera projetinhos pra você desenvolver, alguns são rápidos como um único componente, outros já são projetos completos, é bom pra quem as vezes não consegue ter ideia de algo pra se fazer (como eu).

https://github.com/florinpop17/app-ideas
Tem esse link com diversos projetos para desenvolver mas eu ainda não sei para onde vai ainda. Pois estou no começo dos estudos. Espero que sirva e s servir me adiciona no GitHub quem sabe podemos tentar desenvolver algo juntos!?

**Comment by Fabio-Pina** [comment]
> Sobre sites de desafio, recomendo o [HackerRank](https://www.hackerrank.com/), muita gente usa pra aplicar testes pra seleção. Se não me engano, acertando ou errando acho que dá pra ver a solução depois do problema... não deixa de ser um jeito de aprender praticando. Tem séries que estão formatados muito mais didáticos que se comparados com sites que focam em Coding Challenge.

Ola, no W3C Schools tambem tem isso. Você estuda uma linguagem e aparece uns exercícios. Não sei se é igual esse que você postou! Da uma olhadinha não custa nada.

**Comment by ghost** [comment]
> @leandroruel eu faço essa abordagem também, porém também crio um arquivo .txt e vou pausando o vídeo e anotando o mais detalhado possível o que vou aprendendo para aquilo entrar mais fácil na cabeça, mas não vou mentir que no fundo me sinto fazendo um copy/paste por só ir copiando o que esta sendo ensinado, vou tentar variar o código, ver outras aplicações e etc. para poder praticar um pouco mais. Obrigado pela dica, vou sempre deixar um arquivinho separado fora o que já está rolando na aula pra ir praticando variações o que estou aprendendo.
> 
> @simaodeveloper nem me fale em portfólio, comecei, fiz um basicão e ta lá hahahahaha eu gosto dessa abordagem de projetos, mas aqui fica meu problema com isso e talvez você saiba me ajudar a contornar melhor, você não acaba focando em fazer o projeto apenas com o que já sabe? Você não acaba deixando de estudar coisas novas por estar exercitando uma forma que já conhece?
> Eu vou tentar melhor essa abordagem, porém a cada passo vou ir pesquisando se não tem uma forma melhor de fazer, como nos passos 4 e 5 que você descreveu ai, já vou começar hoje a bolar algum projeto legal pra botar em prática. Obrigado cara.
> 
> Alias dei update na minha issue com um site que gera projetinhos pra você desenvolver, alguns são rápidos como um único componente, outros já são projetos completos, é bom pra quem as vezes não consegue ter ideia de algo pra se fazer (como eu).

Fazer anotações durante as vídeo aulas são bem eficazes. No meu caso percebi que é mais prático fazer anotações com papel e lápis, me ajuda na retenção do conteúdo.

**Comment by ghost** [comment]
É interessante o método de estudo ensinado pelo professor Pierluige Piazzi. Ele escreveu livros que ensinam a técnica de aprendizado circadiano.
